### ISSUE TITLE
bfloat16 numeric stability w/ log_softmax + cross_entropy as based on Flax examples

### ISSUE THREAD

**Issue by rwightman** [google/flax#677]
# Problem you have encountered:

Testing my EfficientNet impl on TPU w/ bfloat16 the training collapsed 2/3 of the way through my training schedule. The models were training reasonably on FP32 GPU, and seemed to be better behaved w/ float16 GPU and the dynamic scaler (although seems the fix here improves late results in that case as well after restoring from an earlier checkpoint). 

I went through my train code (https://github.com/rwightman/efficientnet-jax/blob/master/tf_linen_train.py) based on ImageNet example here, and RMSProp impl (https://github.com/rwightman/efficientnet-jax/blob/master/jeffnet/linen/optim/rmsprop_tensorflow.py) and see any issues.

I ended up fixing the issue by altering the loss fn. Instead of leaving the dtype as bfloat16 for the log_softmax I cast to float32. It is kept in float32 for the onehot and subsequent sum/mean reduction. I don't see this pattern (casting to float32) done in any Flax examples involving log_softmax + xent but it seems like it's something that will crop up for others. I did try leaving log_softmax in bfloat16 and just casting to float32 for the loss reductions. That (may have) reduced the speed of collapse but didn't prevent it.

I've tested the above fix by restoring from a point in my train schedule where the loss was continually increasing and accuracy metrics were falling off the cliff. Everything zipped back to normal. Training from scratch since has been going well.

My loss impl here (https://github.com/rwightman/efficientnet-jax/blob/master/jeffnet/common/loss.py). I keep the softmax outside of my networks.

Another somewhat related issue. I'd already cast to float32 for avgpool means as that was causing issues on GPU mixed-prec. It should be noted that all the ops mentioned here (reductions, log_softmax) are auto cast to fp32 when using AMP in PyTorch on GPU.

### What you expected to happen:

Training to converge to similar results as EfficientNet paper, Google TF TPU impl, and my own PyTorch impl.

### Steps to reproduce:

Run train script linked above. Hparams can be provided.

**Comment by jheek** [comment]
Great to hear you are trying out bfloat16 on TPUs!
I  think you are actually one of the first to try bfloat16 training with EfficientNet. The original implementation of efficientnet used float32 in the whole network. I actually tried to use bfloat16 once in the original implementation and it gave me NaN loses and I am pretty sure other researchers had similar issues.

Actually in prevision resnet implementation we did actually cast to float32 before computing the log_softmax + crossent. But it was later removed because it seemed stable. The log_softmax we use is simply a re-export of `jax.nn.log_softmax` so if we find some instability in its computation (or gradient) we should probably raise this issue with the Jax team.

I'm somewhat surprised about your issue with avgpool because it's really just a (depthwise) convolution with all weights set to a constant.

We normally follow a very similar strategy compared to AMP. That is we do reduction based layers (e.g. BatchNorm) in float32 but store the result in (b)float16. 

I think what we need to resolve this is a test case of just the log_softmax + crossent to see if it can be broken for some inputs when computing in bfloat16. The real issue might be more subtle though. It could be that the numerical error caused by the reduced precision is simply too big for the model to train well. Mixed precision training is more of an art than a science still.

**Comment by rwightman** [comment]
@jheek I have checkpoints around the turning point on one run, I'll make a note to see if I can figure out what the log_softmax input/output distribution looks like as it gets worse.

For the avg pool, I ended up just using mean of spatial dims, not the actual impl since they're all adaptive pools -> 1x1 . For a network life EfficientNet, in terms of the deptwhise conv equivalence, it's a conv you're likely never to see for some cses.. on the large effnets the spatial dims are quite large, unlike a ResNet with one 7x7 avg pool at the end, the effnets go up to 25x25. And then the SqueezeExcite AvgPools are up to 400x400 -> 1x1 near the input on stride 2 features.

I noticed that the BatchNorm is handled properly in terms of casting to float32. But AMP does patch or changes dispatch for all of these ops and more by default vs require selectively include them when stability issues are run into. Both approaches are valid, but this one will likely result in more numeric issues cropping up and may warrant some documentation to raise awareness and help less experienced users debug.
